src/main/java/org/mondemand/Client.java

```diff
@@ -601,12 +601,16 @@ public void increment (StatType type, String key, int value) {
       realKey = ClassUtils.getCallingClass(CALLER_DEPTH);
     }
 
+    // Note: increment could be lost due to a race condition but no
+    //       synchronization is required.
     StatsMessage realValue = (StatsMessage) this.stats.get(realKey);
     if(realValue == null) {
       // create the counter if doesn't exist
-      realValue = new StatsMessage(realKey, type);
-      this.stats.putIfAbsent(realKey, realValue);
-      realValue = this.stats.get(realKey);
+      StatsMessage newValue = new StatsMessage(realKey, type);
+      realValue = this.stats.putIfAbsent(realKey, newValue);
+      if(realValue == null) {
+        realValue = newValue;
+      }
```

Comment:
I think this is the correct pattern.
putIfAbsent returns the existing value, so if it returns non null, then that means another thread had already put and object, else if putIfAbsent returns null, then it means the current thread put the object into the map.